scylla-cql/src/types/deserialize/value.rs

```
@@ -0,0 +1,1515 @@
+/// - `Some(None)` - present, but null
+/// - `Some(Some(...))` - non-null, present value
+pub struct UdtIterator<'frame> {
+    fields: &'frame [(String, ColumnType)],
+    raw_iter: BytesSequenceIterator<'frame>,
+}
+
+impl<'frame> UdtIterator<'frame> {
+    #[inline]
+    pub fn new(fields: &'frame [(String, ColumnType)], slice: FrameSlice<'frame>) -> Self {
+        Self {
+            fields,
+            raw_iter: BytesSequenceIterator::new(slice),
+        }
+    }
+
+    #[inline]
+    pub fn fields(&self) -> &'frame [(String, ColumnType)] {
+        self.fields
+    }
+}
+
+impl<'frame> DeserializeCql<'frame> for UdtIterator<'frame> {
+    fn type_check(typ: &ColumnType) -> Result<(), ParseError> {
+        match typ {
+            ColumnType::UserDefinedType { .. } => Ok(()),
+            _ => Err(ParseError::BadIncomingData(format!(
+                "Expected a user defined type, got {:?}",
+                typ,
+            ))),
+        }
+    }
+
+    fn deserialize(
+        typ: &'frame ColumnType,
+        v: Option<FrameSlice<'frame>>,
+    ) -> Result<Self, ParseError> {
+        let v = ensure_not_null_slice(v)?;
+        let mem = v.as_slice();
+        let fields = match typ {
+            ColumnType::UserDefinedType { field_types, .. } => field_types.as_ref(),
+            _ => {
+                return Err(ParseError::BadIncomingData(format!(
+                    "Expected a user defined type, got {:?}",
+                    typ,
+                )))
+            }
+        };
+        Ok(Self::new(
+            fields,
+            FrameSlice::new_subslice(mem, v.as_bytes_ref()),
+        ))
+    }
+}
+
+impl<'frame> Iterator for UdtIterator<'frame> {
+    type Item = Result<
+        (
+            &'frame (String, ColumnType),
+            Option<Option<FrameSlice<'frame>>>,
+        ),
+        ParseError,
+    >;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // TODO: Should we fail when there are too many fields?
```

Let's discuss this. Should we?

If we got more fields than we expected for a UDT, why should we let it silently pass? IMO it deserves at least a warning. Failure is appropriate as well, as I can't find any counterexample.

---

The protocolspec only allows less fields to be sent, and the description implies that it's incorrect to send more:

```
  A UDT value is composed of successive [bytes] values, one for each field of the UDT
  value (in the order defined by the type). A UDT value will generally have one value
  for each field of the type it represents, but it is allowed to have less values than
  the type has fields.
```

The TODO is mostly about considering how to handle a case where the database sends a response that does not conform to the spec. If we don't trust the DB then this method will have to be written a bit differently.